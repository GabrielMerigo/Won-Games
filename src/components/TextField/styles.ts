
import styled, { css, DefaultTheme } from 'styled-components';
import { TextFieldProps } from './'

const wrapperModifers = {
  left: () => css`
    left: 0;
    margin-left: 10px;
  `,
  right: () => css`
    right: 0;
    margin-right: 15px;
  `,
  disabled: (theme: DefaultTheme) => css`
    ${Label},
    ${Input},
    ${Icon}{
      cursor: not-allowed;
      color: ${theme.colors.gray};

      &::placeholder {
        color: currentTarget;
      }
    }
  `,
  error: (theme: DefaultTheme) => css`
   ${InputWrapper}{
     border-color: ${theme.colors.red};
   }

   ${Icon},
   ${Label}{
      color: ${theme.colors.red};
   }
  `
}

type IconPositionProps = { iconPosition: 'right' | 'left' };
type WrapperProps = Pick<TextFieldProps, 'disabled'> & { error: boolean }

export const InputWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    background: ${theme.colors.lightGray};
    border-radius: 0.2rem;
    border: 0.2rem solid;
    border-color: ${theme.colors.lightGray};
    position: relative;

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }
  `}
`

export const Input = styled.input<IconPositionProps>`
  ${({ theme, iconPosition }) => css`
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} 0;
    background: transparent;
    border: 0;
    outline: none;
    width: 100%;

    ${iconPosition === 'right'
      ? css`margin-right: 15px;`
      : css`margin-left: 15px;`
    };
  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.black};
    cursor: pointer;
  `}
`;

export const Icon = styled.div<IconPositionProps>`
  ${({ theme, iconPosition }) => css`
    width: 3rem;
    color: ${theme.colors.gray};
    margin-left: ${theme.spacings.xxsmall};
    margin-top: 0.4rem;
    position: absolute;

    ${iconPosition === 'left' && wrapperModifers.left()}
    ${iconPosition === 'right' && wrapperModifers.right()}

    & > svg {
      width: 100%;
    }
  `}
`;

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, disabled, error }) => css`
    ${error && wrapperModifers.error(theme)}
    ${disabled && wrapperModifers.disabled(theme)}
  `}
`;

export const Error = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.red};
    font-size: ${theme.font.sizes.xsmall};
  `}
`
