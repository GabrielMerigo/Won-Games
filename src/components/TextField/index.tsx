import { useState, InputHTMLAttributes } from 'react'

import * as S from './styles'

export type TextFieldProps = {
  onInput?: (value: string) => void
  label?: string
  name?: string
  initialValue?: string
  icon?: JSX.Element,
  iconPosition?: 'right' | 'left'
  error?: string
} & InputHTMLAttributes<HTMLInputElement>

const TextField = ({
  label,
  initialValue = '',
  name,
  onInput,
  icon,
  iconPosition = 'left',
  disabled = false,
  error = '',
  ...props
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue)

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = e.currentTarget.value
    setValue(newValue)

    !!onInput && onInput(newValue)
  }

  return (
    <S.Wrapper disabled={disabled} error={!!error}>
      {!!label && <S.Label htmlFor={name}>{label}</S.Label>}
      <S.InputWrapper>
        {!!icon && <S.Icon iconPosition={iconPosition}>{icon}</S.Icon>}
        <S.Input
          iconPosition={iconPosition}
          type="text"
          onChange={onChange}
          value={value}
          disabled={disabled}
          name={name}
          {...(label ? { id: name } : {})}
          {...props}
        />
        <S.Error>{error}</S.Error>
      </S.InputWrapper>
    </S.Wrapper>
  )
}

export default TextField
