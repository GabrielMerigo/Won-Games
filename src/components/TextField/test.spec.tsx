
import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Email } from '@styled-icons/material-outlined/Email';

import { renderWithTheme } from 'utils/tests/helpers'

import TextField from '.'
import { render } from 'react-dom';

describe('<TextField />', () => {
  it('Renders with Label', () => {
    renderWithTheme(<TextField label="Label" name="Field" id="Field" />)

    expect(screen.getByLabelText('Label')).toBeInTheDocument()
  })

  it('Renders without Label', () => {
    renderWithTheme(<TextField />)

    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument()
  })

  it('Renders with placeholder', () => {
    renderWithTheme(<TextField placeholder="hey you" />)
    expect(screen.getByPlaceholderText('hey you')).toBeInTheDocument()
  })

  it('Changes its value when typing', async () => {
    const onInput = jest.fn()
    renderWithTheme(
      <TextField
        onInput={onInput}
        label="TextField"
        name="TextField"
        id="TextField"
      />
    );

    const input = screen.getByRole('textbox');
    const text = 'This is my new text';
    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).toHaveValue(text);
      expect(onInput).toHaveBeenCalledTimes(text.length);
    });

    expect(onInput).toHaveBeenCalledWith(text);
  })

  it('Is accessible by tab', () => {
    renderWithTheme(
      <TextField label="TextField" name="TextField" />
    );

    const input = screen.getByLabelText('TextField');
    expect(document.body).toHaveFocus();

    userEvent.tab();
    expect(input).toHaveFocus();
  })

  it('should render with icon', () => {
    renderWithTheme(
      <TextField icon={<Email data-testid="icon" />} />
    );

    expect(screen.getByTestId('icon')).toBeInTheDocument();
  });

  it('should render icon on left side by default ', () => {
    renderWithTheme(
      <TextField id="Field" icon={<Email data-testid="icon" />} />
    );

    expect(screen.getByTestId(/icon/i).parentElement).toHaveStyle({
      left: 0,
      'margin-left': '10px'
    });

    expect(screen.getByRole('textbox')).toHaveStyle({
      'margin-left': '30px'
    });
  })

  it('should render icon on right side ', () => {
    renderWithTheme(
      <TextField iconPosition="right" id="Field" icon={<Email data-testid="icon" />} />
    );

    expect(screen.getByTestId(/icon/i).parentElement).toHaveStyle({
      right: 0,
      'margin-right': '15px'
    });

    expect(screen.getByRole('textbox')).toHaveStyle({
      'margin-right': '30px'
    });
  })


  it("shouldn't use button when the one is disabled", async () => {
    const onInput = jest.fn();

    renderWithTheme(
      <TextField
        onInput={onInput}
        label="TextField"
        name="TextField"
        id="TextField"
        disabled
      />
    );

    const input = screen.getByRole('textbox');
    expect(input).toBeDisabled();

    const text = 'This is my new Text';
    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).not.toHaveValue();
    });

    expect(onInput).not.toHaveBeenCalled();
  })


  it('Renders with error', () => {
    const onInput = jest.fn();

    const { container } = renderWithTheme(<TextField
      label="TextField"
      name="TextField"
      id="TextField"
      error="error"
    />)


    expect(screen.getByText('error')).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  })
})
