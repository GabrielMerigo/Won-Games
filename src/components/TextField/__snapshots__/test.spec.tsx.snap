// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> Renders with error 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid;
  border-color: #EAEAEA;
  position: relative;
}

.c3:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c4 {
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
  margin-left: 30px;
}

.c1 {
  font-size: 1.4rem;
  color: #030517;
  cursor: pointer;
}

.c0 .c2 {
  border-color: #FF6347;
}

.c5 {
  color: #FF6347;
  font-size: 1.2rem;
}

<div
  class="c0"
>
  <label
    class="c1"
    for="TextField"
  >
    TextField
  </label>
  <div
    class="c2 c3"
  >
    <input
      class="c4"
      id="TextField"
      type="text"
      value=""
    />
    <p
      class="c5"
    >
      error
    </p>
  </div>
</div>
`;
